src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -2436,118 +2476,312 @@ public void UpdateIsNullableEnabledForConstructorsAndFields(bool useStatic, bool
+                this.IsNullableEnabledForInstanceConstructorsAndFields = isNullableEnabledForInstanceConstructorsAndFields;
+                this.IsNullableEnabledForStaticConstructorsAndFields = isNullableEnabledForStaticConstructorsAndFields;
+            }
+        }
+
+        private sealed class MembersAndInitializersBuilder
+        {
+            public ArrayBuilder<Symbol> NonTypeNonIndexerMembers;
+            private ArrayBuilder<FieldOrPropertyInitializer>? InstanceInitializersForPositionalMembers;
+            private bool IsNullableEnabledForInstanceConstructorsAndFields;
+            private bool IsNullableEnabledForStaticConstructorsAndFields;
+
+            public MembersAndInitializersBuilder(DeclaredMembersAndInitializers declaredMembersAndInitializers)
+                : this(
+                    declaredMembersAndInitializers.NonTypeNonIndexerMembers,
+                    declaredMembersAndInitializers.IsNullableEnabledForInstanceConstructorsAndFields,
+                    declaredMembersAndInitializers.IsNullableEnabledForStaticConstructorsAndFields)
+            {
+            }
+
+            private MembersAndInitializersBuilder(
+                ImmutableArray<Symbol> nonTypeNonIndexerMembers,
+                bool isNullableEnabledForInstanceConstructorsAndFields,
+                bool isNullableEnabledForStaticConstructorsAndFields)
+            {
+                Debug.Assert(!nonTypeNonIndexerMembers.IsDefault);
+
+                this.NonTypeNonIndexerMembers = getAndFillBuilder(nonTypeNonIndexerMembers);
+
+                this.IsNullableEnabledForInstanceConstructorsAndFields = isNullableEnabledForInstanceConstructorsAndFields;
+                this.IsNullableEnabledForStaticConstructorsAndFields = isNullableEnabledForStaticConstructorsAndFields;
+
+                static ArrayBuilder<T> getAndFillBuilder<T>(ImmutableArray<T> original)
                 {
-                    group.Free();
-                    return ImmutableArray<FieldOrPropertyInitializer>.Empty;
+                    var result = ArrayBuilder<T>.GetInstance(original.Length);
+                    result.AddRange(original);
+                    return result;
                 }
+            }
 
-                var builder = ArrayBuilder<FieldOrPropertyInitializer>.GetInstance(group.Count);
+            public MembersAndInitializers ToReadOnlyAndFree(DeclaredMembersAndInitializers declaredMembers)
+            {
+                var instanceInitializers = InstanceInitializersForPositionalMembers is null
+                    ? declaredMembers.InstanceInitializers
+                    : mergeInitializers();
 
-                foreach (FieldOrPropertyInitializer.Builder initializer in group)
-                {
-                    builder.Add(new FieldOrPropertyInitializer(initializer.FieldOpt, initializer.Syntax, syntaxLength));
+                return new MembersAndInitializers(
+                    NonTypeNonIndexerMembers.ToImmutableAndFree(),
```

Comment:
>NonTypeNonIndexerMembers.ToImmutableAndFree() [](start = 20, length = 45)

I think we should reuse existing immutable array if we didn't add any new members.